### PR TITLE
fix: table selectable callback index incorrect issue

### PR DESCRIPTION
**Pull request by msidolphin**
Please make sure these boxes are checked before submitting your PR, thank you!

* [x] Make sure you follow Element's contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
* [x] Make sure you are merging your commits to `dev` branch.
* [x] Add some descriptions and refer to relative issues for your PR.

### CODE REVIEW

**Comment by JeremyWuuuuu** [comment]
Should we add test cases for verifying?

**Comment by msidolphin** [comment]
> Should we add test cases for verifying?

I already added test case

File: packages/table/__tests__/table-column.spec.ts [hunk 1]
```diff
@@ -944,4 +944,139 @@ describe('table column', () => {
...
+      const wrapper = createTable({
+        selectable(row, index) {
+          expect(row.index - 1).toEqual(index)
+          return !row.disabled
+        },
+      })
+      await nextTick()
+      wrapper.vm.$refs.table.toggleAllSelection()
+      wrapper.unmount()
+    })
```
Comment (JeremyWuuuuu) [inline packages/table/__tests__/table-column.spec.ts:1079]: use assertion like this:

```ts
const expectToBeCalled = jest.fn()
const wrapper = createTable({
        selectable(row, index) {
          expectToBeCalled((row.index - 1) === index)
          return !row.disabled
        },
      })
await nextTick()
wrapper.vm.$refs.table.toggleAllSelection()

expect(expectToBeCalled).toHaveBeenCalledWith(true)

```

Comment (msidolphin) [reply packages/table/__tests__/table-column.spec.ts]: I need to make sure that the result is correct every time, but I refactored the code anyway

Comment (JeremyWuuuuu) [reply packages/table/__tests__/table-column.spec.ts]: Humm, I think you can do that by asserting your result to `expect(toBeCalled).toHaveBeenNthCalled(order, params)`. Use a for loop to assert your result

File: packages/table/__tests__/table-column.spec.ts [hunk 2]
```diff
@@ -944,4 +944,141 @@ describe('table column', () => {
...
+      const result = []
+      const wrapper = createTable({
+        selectable(row, index) {
+          result.push((row.index - 1) === index)
+          return !row.disabled
+        },
+      })
+      await nextTick()
+      wrapper.vm.$refs.table.toggleAllSelection()
+      expect(result.every(item => item)).toBeTruthy()
```
Comment (JeremyWuuuuu) [inline packages/table/__tests__/table-column.spec.ts:1079]: This is a way to test this case. Nice work.

**Review by jw-foss — APPROVED** [review]

**Pull request merged**